File: app/routers/platforms.py
import json
import os
import base64
from datetime import datetime, timedelta, timezone
from typing import Any

import httpx
from fastapi import APIRouter, Depends, HTTPException, Request, status
from fastapi.responses import JSONResponse, RedirectResponse
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import settings
from app.dependencies import get_current_user, get_db, get_session_user_id
from app.models.platform_account import PlatformAccount
from app.models.user import User
from app.schemas.platform import LemmyConnectRequest, MastodonConnectRequest, PlatformAccountOut
from app.services.crypto import decrypt, encrypt
from app.services.platforms.lemmy import LemmyClient
from app.services.platforms.linkedin import LinkedInClient
from app.services.platforms.youtube import YouTubeClient

router = APIRouter(prefix="/api/platforms", tags=["platforms"])

LINKEDIN_AUTH_URL = "https://www.linkedin.com/oauth/v2/authorization"
LINKEDIN_TOKEN_URL = "https://www.linkedin.com/oauth/v2/accessToken"
LINKEDIN_REDIRECT_URI = "http://localhost:8000/api/platforms/linkedin/callback"

YOUTUBE_AUTH_URL = "https://accounts.google.com/o/oauth2/v2/auth"
YOUTUBE_TOKEN_URL = "https://oauth2.googleapis.com/token"
YOUTUBE_SCOPES = "https://www.googleapis.com/auth/youtube.upload https://www.googleapis.com/auth/youtube.force-ssl https://www.googleapis.com/auth/yt-analytics.readonly"


@router.get("", response_model=list[PlatformAccountOut])
async def list_platforms(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(PlatformAccount).where(PlatformAccount.user_id == user.id)
    )
    return result.scalars().all()


@router.delete("/{account_id}", status_code=status.HTTP_204_NO_CONTENT)
async def disconnect_platform(
    account_id: int,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(PlatformAccount).where(
            PlatformAccount.id == account_id, PlatformAccount.user_id == user.id
        )
    )
    account = result.scalar_one_or_none()
    if not account:
        raise HTTPException(status_code=404, detail="Platform account not found")
    await db.delete(account)
    await db.commit()


# ── Mastodon ──────────────────────────────────────────────────────────────────

@router.post("/mastodon/connect")
async def mastodon_connect(
    body: MastodonConnectRequest,
    request: Request,
    user: User = Depends(get_current_user),
):
    instance = body.instance_url.rstrip("/")
    # Register the app with this instance
    async with httpx.AsyncClient() as client:
        reg = await client.post(
            f"{instance}/api/v1/apps",
            json={
                "client_name": "Markdept",
                "redirect_uris": "http://localhost:8000/api/platforms/mastodon/callback",
                "scopes": "read write",
                "website": "http://localhost:8000",
            },
        )
        reg.raise_for_status()
        app_data = reg.json()

    client_id = app_data["client_id"]
    client_secret = app_data["client_secret"]
    state = base64.urlsafe_b64encode(os.urandom(16)).rstrip(b"=").decode()

    # Store temp data in a cookie
    temp = json.dumps(
        {"instance": instance, "client_id": client_id, "client_secret": client_secret, "state": state}
    )
    auth_url = (
        f"{instance}/oauth/authorize"
        f"?response_type=code"
        f"&client_id={client_id}"
        f"&redirect_uri=http://localhost:8000/api/platforms/mastodon/callback"
        f"&scope=read+write"
        f"&state={state}"
    )
    response = JSONResponse({"auth_url": auth_url})
    response.set_cookie("mastodon_connect", temp, max_age=600, httponly=True, samesite="lax")
    return response


@router.get("/mastodon/callback")
async def mastodon_callback(
    request: Request,
    code: str,
    state: str | None = None,
    db: AsyncSession = Depends(get_db),
):
    user_id = get_session_user_id(request)
    if not user_id:
        raise HTTPException(status_code=401, detail="Not authenticated")

    temp_raw = request.cookies.get("mastodon_connect")
    if not temp_raw:
        raise HTTPException(status_code=400, detail="Missing Mastodon connect session")
    temp = json.loads(temp_raw)
    instance = temp["instance"]

    async with httpx.AsyncClient() as client:
        resp = await client.post(
            f"{instance}/oauth/token",
            json={
                "grant_type": "authorization_code",
                "code": code,
                "client_id": temp["client_id"],
                "client_secret": temp["client_secret"],
                "redirect_uri": "http://localhost:8000/api/platforms/mastodon/callback",
                "scope": "read write",
            },
        )
        resp.raise_for_status()
        token_data = resp.json()
        access_token = token_data["access_token"]

        # Get account info
        me_resp = await client.get(
            f"{instance}/api/v1/accounts/verify_credentials",
            headers={"Authorization": f"Bearer {access_token}"},
        )
        me_resp.raise_for_status()
        me = me_resp.json()

    account = PlatformAccount(
        user_id=user_id,
        platform="mastodon",
        account_name=f"@{me['username']}@{instance.replace('https://', '')}",
        account_id_on_platform=me["id"],
        instance_url=instance,
        access_token_encrypted=encrypt(access_token),
    )
    db.add(account)
    await db.commit()

    response = RedirectResponse(url="/platforms")
    response.delete_cookie("mastodon_connect")
    return response


# ── LinkedIn ──────────────────────────────────────────────────────────────────

@router.post("/linkedin/connect")
async def linkedin_connect(user: User = Depends(get_current_user)):
    state = base64.urlsafe_b64encode(os.urandom(16)).rstrip(b"=").decode()
    auth_url = (
        f"{LINKEDIN_AUTH_URL}"
        f"?response_type=code"
        f"&client_id={settings.LINKEDIN_CLIENT_ID}"
        f"&redirect_uri={LINKEDIN_REDIRECT_URI}"
        f"&scope=r_liteprofile%20r_emailaddress%20w_member_social"
        f"&state={state}"
    )
    return RedirectResponse(url=auth_url)


@router.get("/linkedin/connect")
async def linkedin_connect_get(user: User = Depends(get_current_user)):
    state = base64.urlsafe_b64encode(os.urandom(16)).rstrip(b"=").decode()
    auth_url = (
        f"{LINKEDIN_AUTH_URL}"
        f"?response_type=code"
        f"&client_id={settings.LINKEDIN_CLIENT_ID}"
        f"&redirect_uri={LINKEDIN_REDIRECT_URI}"
        f"&scope=r_liteprofile%20r_emailaddress%20w_member_social"
        f"&state={state}"
    )
    return RedirectResponse(url=auth_url)


@router.get("/linkedin/callback")
async def linkedin_callback(
    request: Request,
    code: str,
    state: str | None = None,
    db: AsyncSession = Depends(get_db),
):
    user_id = get_session_user_id(request)
    if not user_id:
        raise HTTPException(status_code=401, detail="Not authenticated")

    async with httpx.AsyncClient() as client:
        resp = await client.post(
            LINKEDIN_TOKEN_URL,
            data={
                "grant_type": "authorization_code",
                "code": code,
                "redirect_uri": LINKEDIN_REDIRECT_URI,
                "client_id": settings.LINKEDIN_CLIENT_ID,
                "client_secret": settings.LINKEDIN_CLIENT_SECRET,
            },
            headers={"Content-Type": "application/x-www-form-urlencoded"},
        )
        resp.raise_for_status()
        token_data = resp.json()
        access_token = token_data["access_token"]
        expires_in = token_data.get("expires_in", 5184000)

    li_client = LinkedInClient()
    urn = await li_client.get_profile_urn(access_token)
    account_name = urn.split(":")[-1]

    expires_at = datetime.now(timezone.utc) + timedelta(seconds=expires_in)
    account = PlatformAccount(
        user_id=user_id,
        platform="linkedin",
        account_name=account_name,
        account_id_on_platform=urn,
        access_token_encrypted=encrypt(access_token),
        token_expires_at=expires_at,
    )
    db.add(account)
    await db.commit()
    return RedirectResponse(url="/platforms")


# ── Lemmy ─────────────────────────────────────────────────────────────────────

@router.post("/lemmy/connect")
async def lemmy_connect(
    body: LemmyConnectRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    try:
        jwt, account_name, extra = await LemmyClient.connect(
            body.instance_url, body.username, body.password
        )
    except httpx.HTTPStatusError as e:
        raise HTTPException(status_code=400, detail=f"Lemmy login failed: {e.response.text}")

    account = PlatformAccount(
        user_id=user.id,
        platform="lemmy",
        account_name=account_name,
        account_id_on_platform=extra.get("account_id", ""),
        instance_url=body.instance_url,
        access_token_encrypted=encrypt(jwt),
        extra_data_encrypted=encrypt(json.dumps(extra)),
    )
    db.add(account)
    await db.commit()
    return {"status": "connected", "account_name": account_name}


# ── YouTube ───────────────────────────────────────────────────────────────────

@router.get("/youtube/connect")
async def youtube_connect(user: User = Depends(get_current_user)):
    state = base64.urlsafe_b64encode(os.urandom(16)).rstrip(b"=").decode()
    auth_url = (
        f"{YOUTUBE_AUTH_URL}"
        f"?response_type=code"
        f"&client_id={settings.YOUTUBE_CLIENT_ID}"
        f"&redirect_uri={settings.YOUTUBE_REDIRECT_URI}"
        f"&scope={YOUTUBE_SCOPES.replace(' ', '%20')}"
        f"&access_type=offline"
        f"&prompt=consent"
        f"&state={state}"
    )
    return RedirectResponse(url=auth_url)


@router.post("/youtube/connect")
async def youtube_connect_post(user: User = Depends(get_current_user)):
    return await youtube_connect(user)


@router.get("/youtube/callback")
async def youtube_callback(
    request: Request,
    code: str,
    state: str | None = None,
    db: AsyncSession = Depends(get_db),
):
    user_id = get_session_user_id(request)
    if not user_id:
        raise HTTPException(status_code=401, detail="Not authenticated")

    async with httpx.AsyncClient() as client:
        resp = await client.post(
            YOUTUBE_TOKEN_URL,
            data={
                "grant_type": "authorization_code",
                "code": code,
                "redirect_uri": settings.YOUTUBE_REDIRECT_URI,
                "client_id": settings.YOUTUBE_CLIENT_ID,
                "client_secret": settings.YOUTUBE_CLIENT_SECRET,
            },
        )
        resp.raise_for_status()
        token_data = resp.json()
        access_token = token_data["access_token"]
        refresh_token = token_data.get("refresh_token")
        expires_in = token_data.get("expires_in", 3600)

    yt_client = YouTubeClient()
    channel_id = await yt_client.get_channel_id(access_token)
    expires_at = datetime.now(timezone.utc) + timedelta(seconds=expires_in)

    account = PlatformAccount(
        user_id=user_id,
        platform="youtube",
        account_name=channel_id or "YouTube Channel",
        account_id_on_platform=channel_id,
        access_token_encrypted=encrypt(access_token),
        refresh_token_encrypted=encrypt(refresh_token) if refresh_token else None,
        token_expires_at=expires_at,
    )
    db.add(account)
    await db.commit()
    return RedirectResponse(url="/platforms")

File: src/routes/platforms.ts
import { Hono } from "hono";
import { and, eq } from "drizzle-orm";
import { db } from "../db";
import { platformAccounts, users } from "../db/schema";
import { requireAuth } from "../auth/middleware";
import { encrypt, decrypt } from "../services/crypto";
import { config } from "../config";

const router = new Hono();
router.use("*", requireAuth);

router.get("/", async (c) => {
  const user = c.get("user" as never) as typeof users.$inferSelect;
  const rows = await db.select().from(platformAccounts).where(eq(platformAccounts.userId, user.id));
  return c.json(rows.map(({ accessTokenEncrypted, refreshTokenEncrypted, extraDataEncrypted, ...safe }) => safe));
});

router.delete("/:id", async (c) => {
  const user = c.get("user" as never) as typeof users.$inferSelect;
  const id = Number(c.req.param("id"));
  const rows = await db.delete(platformAccounts).where(and(eq(platformAccounts.id, id), eq(platformAccounts.userId, user.id))).returning();
  if (!rows[0]) return c.json({ detail: "Not found" }, 404);
  return c.json({ ok: true });
});

// Mastodon OAuth flow
router.get("/mastodon/connect", async (c) => {
  const params = new URLSearchParams({
    client_id: config.MASTODON_CLIENT_ID,
    scope: "read write",
    redirect_uri: `${new URL(config.ANTHROPIC_REDIRECT_URI).origin}/api/platforms/mastodon/callback`,
    response_type: "code",
  });
  return c.redirect(`${config.MASTODON_INSTANCE_URL}/oauth/authorize?${params}`);
});

router.get("/mastodon/callback", async (c) => {
  const user = c.get("user" as never) as typeof users.$inferSelect;
  const code = c.req.query("code");
  if (!code) return c.redirect("/platforms?error=mastodon_auth_failed");

  const tokenRes = await fetch(`${config.MASTODON_INSTANCE_URL}/oauth/token`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      client_id: config.MASTODON_CLIENT_ID,
      client_secret: config.MASTODON_CLIENT_SECRET,
      redirect_uri: `${new URL(config.ANTHROPIC_REDIRECT_URI).origin}/api/platforms/mastodon/callback`,
      grant_type: "authorization_code",
      code,
      scope: "read write",
    }),
  });
  if (!tokenRes.ok) return c.redirect("/platforms?error=mastodon_token_failed");
  const tokenData = await tokenRes.json() as { access_token: string };

  const verifyRes = await fetch(`${config.MASTODON_INSTANCE_URL}/api/v1/accounts/verify_credentials`, {
    headers: { Authorization: `Bearer ${tokenData.access_token}` },
  });
  const accountInfo = await verifyRes.json() as { id: string; username: string };

  await db.insert(platformAccounts).values({
    userId: user.id,
    platform: "mastodon",
    accountName: accountInfo.username,
    accountId: accountInfo.id,
    accessTokenEncrypted: await encrypt(tokenData.access_token),
  }).onConflictDoNothing();
  return c.redirect("/platforms?connected=mastodon");
});

// LinkedIn OAuth flow
router.get("/linkedin/connect", async (c) => {
  const state = btoa(String.fromCharCode(...crypto.getRandomValues(new Uint8Array(16))));
  const params = new URLSearchParams({
    response_type: "code",
    client_id: config.LINKEDIN_CLIENT_ID,
    redirect_uri: `${new URL(config.ANTHROPIC_REDIRECT_URI).origin}/api/platforms/linkedin/callback`,
    scope: "openid profile email w_member_social",
    state,
  });
  return c.redirect(`https://www.linkedin.com/oauth/v2/authorization?${params}`);
});

router.get("/linkedin/callback", async (c) => {
  const user = c.get("user" as never) as typeof users.$inferSelect;
  const code = c.req.query("code");
  if (!code) return c.redirect("/platforms?error=linkedin_auth_failed");

  const tokenRes = await fetch("https://www.linkedin.com/oauth/v2/accessToken", {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      grant_type: "authorization_code",
      code,
      redirect_uri: `${new URL(config.ANTHROPIC_REDIRECT_URI).origin}/api/platforms/linkedin/callback`,
      client_id: config.LINKEDIN_CLIENT_ID,
      client_secret: config.LINKEDIN_CLIENT_SECRET,
    }),
  });
  if (!tokenRes.ok) return c.redirect("/platforms?error=linkedin_token_failed");
  const tokenData = await tokenRes.json() as { access_token: string; expires_in?: number };

  const profileRes = await fetch("https://api.linkedin.com/v2/me", {
    headers: { Authorization: `Bearer ${tokenData.access_token}` },
  });
  const profile = await profileRes.json() as { id: string; localizedFirstName?: string; localizedLastName?: string };
  const name = `${profile.localizedFirstName ?? ""} ${profile.localizedLastName ?? ""}`.trim() || profile.id;

  const expiresAt = tokenData.expires_in ? new Date(Date.now() + tokenData.expires_in * 1000) : null;
  await db.insert(platformAccounts).values({
    userId: user.id,
    platform: "linkedin",
    accountName: name,
    accountId: profile.id,
    accessTokenEncrypted: await encrypt(tokenData.access_token),
    tokenExpiresAt: expiresAt,
  }).onConflictDoNothing();
  return c.redirect("/platforms?connected=linkedin");
});

// Lemmy — manual credential setup
router.post("/lemmy/connect", async (c) => {
  const user = c.get("user" as never) as typeof users.$inferSelect;
  const body = await c.req.json() as { username: string; password: string; communityId: number };
  const extra = JSON.stringify({ username: body.username, password: body.password, communityId: body.communityId });
  await db.insert(platformAccounts).values({
    userId: user.id,
    platform: "lemmy",
    accountName: body.username,
    extraDataEncrypted: await encrypt(extra),
  }).onConflictDoNothing();
  return c.json({ ok: true }, 201);
});

// YouTube OAuth flow
router.get("/youtube/connect", async (c) => {
  const state = btoa(String.fromCharCode(...crypto.getRandomValues(new Uint8Array(16))));
  const params = new URLSearchParams({
    client_id: config.YOUTUBE_CLIENT_ID,
    redirect_uri: config.YOUTUBE_REDIRECT_URI,
    response_type: "code",
    scope: "https://www.googleapis.com/auth/youtube.upload https://www.googleapis.com/auth/youtube.readonly",
    access_type: "offline",
    prompt: "consent",
    state,
  });
  return c.redirect(`https://accounts.google.com/o/oauth2/v2/auth?${params}`);
});

router.get("/youtube/callback", async (c) => {
  const user = c.get("user" as never) as typeof users.$inferSelect;
  const code = c.req.query("code");
  if (!code) return c.redirect("/platforms?error=youtube_auth_failed");

  const tokenRes = await fetch("https://oauth2.googleapis.com/token", {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      code,
      client_id: config.YOUTUBE_CLIENT_ID,
      client_secret: config.YOUTUBE_CLIENT_SECRET,
      redirect_uri: config.YOUTUBE_REDIRECT_URI,
      grant_type: "authorization_code",
    }),
  });
  if (!tokenRes.ok) return c.redirect("/platforms?error=youtube_token_failed");
  const tokenData = await tokenRes.json() as { access_token: string; refresh_token: string; expires_in?: number };

  const channelRes = await fetch("https://www.googleapis.com/youtube/v3/channels?part=snippet&mine=true", {
    headers: { Authorization: `Bearer ${tokenData.access_token}` },
  });
  const channelData = await channelRes.json() as { items?: Array<{ id: string; snippet: { title: string } }> };
  const channel = channelData.items?.[0];

  const expiresAt = tokenData.expires_in ? new Date(Date.now() + tokenData.expires_in * 1000) : null;
  await db.insert(platformAccounts).values({
    userId: user.id,
    platform: "youtube",
    accountName: channel?.snippet.title ?? "YouTube Channel",
    accountId: channel?.id ?? "",
    accessTokenEncrypted: await encrypt(tokenData.access_token),
    refreshTokenEncrypted: await encrypt(tokenData.refresh_token),
    tokenExpiresAt: expiresAt,
  }).onConflictDoNothing();
  return c.redirect("/platforms?connected=youtube");
});

export default router;
